rbc/stdlib/creation_functions.py

```diff
@@ -4,21 +4,38 @@
 https://data-apis.org/array-api/latest/API_specification/creation_functions.html
 """
 
-from rbc import typesystem
+from rbc import typesystem, errors
 from rbc.omnisci_backend.omnisci_array import Array, ArrayPointer
-from rbc.stdlib import Expose
-from numba import njit
+from rbc.stdlib import Expose, API
 from numba.core import extending, types
+from numba.core.typing import asnumbatype
 
 __all__ = [
     'full', 'full_like', 'empty_like', 'empty', 'zeros', 'zeros_like',
-    'ones', 'ones_like', 'array', 'cumsum'
+    'ones', 'ones_like', 'cumsum'
 ]
 
 
 expose = Expose(globals(), 'creation_functions')
 
 
+def _determine_dtype(dtype, fill_value):
+    if dtype is None:
+        # determine from fill_value
+        if fill_value is None:
+            # return the default floating-point data type
+            return types.double
+        else:
+            return asnumbatype.typeof(fill_value)
+    else:
+        if isinstance(dtype, types.UnicodeType):
+            raise errors.RequireLiteralValue(dtype)
+        elif isinstance(dtype, types.StringLiteral):
+            return typesystem.Type.fromstring(dtype.literal_value).tonumba()
+        else:
+            return typesystem.Type.fromobject(dtype).tonumba()
```

Comment:
Here `dtype` can be any object, right? For instance, a user could call `array_api.zeros(3, dtype=123)` and it will fail at calling `typesystem.Type.fromobject`.

I suggest that withing `_determine_dtype` we explicitly check the correctness of `dtype` argument type: it can be either `str` object or `type` object, all other types (ints, floats, lists, etc) should raise a TypeError (that, if possible, would circumvent the verbose numba exception handling).